amaranth/hdl/_ir.py

```diff
@@ -3,17 +3,23 @@
 import enum
 import warnings
 
-from .._utils import flatten, to_binary
+from .._utils import flatten, to_binary, final
 from .. import tracer, _unused
 from . import _ast, _cd, _ir, _nir
 
 
 __all__ = [
-    "UnusedElaboratable", "Elaboratable", "DuplicateElaboratable", "DriverConflict", "Fragment",
-    "Instance", "IOBufferInstance", "PortDirection", "Design", "build_netlist",
+    "AlreadyElaborated", "UnusedElaboratable", "Elaboratable", "DuplicateElaboratable",
+    "DriverConflict", "Fragment", "Instance", "IOBufferInstance", "PortDirection", "Design",
+    "build_netlist",
 ]
 
 
+@final
+class AlreadyElaborated(Exception):
+    """Exception raised when a module is being modified after elaboration."""
```

Comment:
A memory is not a module.

```suggestion
    """Exception raised when an elaboratable is being modified after elaboration."""
```